platform/wayland/cog-platform-wl.c

```diff
@@ -281,22 +339,248 @@ static struct {
+    .configure = shell_popup_surface_configure,
+    .popup_done = shell_popup_surface_popup_done,
+};
+
+static void
+create_wl_shell_popup()
+{
+    popup_data.shell_surface =
+        wl_shell_get_shell_surface(wl_data.shell, popup_data.shell_context->wl_shell_data.wl_surface);
+    g_assert(popup_data.shell_surface);
+
+    wl_shell_surface_add_listener(popup_data.shell_surface, &shell_popup_surface_listener, NULL);
+    wl_shell_surface_set_popup(popup_data.shell_surface, wl_data.seat, wl_data.event_serial,
+                               win_data.shell_context->wl_shell_data.wl_surface, 0,
+                               (win_data.height - popup_data.height), 0);
+
+    display_popup();
+}
+
+static shell_functions wayland_shell_functions = {.enter_fullscreen = &wl_shell_enter_fullscreen,
+                                                  .exit_fullscreen = &wl_shell_exit_fullscreen,
+                                                  .destroy_shell = &destroy_wl_shell,
+                                                  .maximize_surface = &maximize_wl_shell_surface,
+                                                  .create_window = &create_wl_shell_window,
+                                                  .create_popup = &create_wl_shell_popup};
+
+wl_shell_data wayland_data = {.functions = &wayland_shell_functions};
+
+// f shell operations
+static void
+f_shell_enter_fullscreen()
+{
+    win_data.should_resize_to_largest_output = true;
+    resize_to_largest_output();
+}
+static void
+f_shell_exit_fullscreen()
+{
+    configure_surface_geometry(win_data.width_before_fullscreen, win_data.height_before_fullscreen);
+    resize_window();
+}
+static void
+destroy_f_shell()
+{
+    zwp_fullscreen_shell_v1_destroy(wl_data.fshell);
+}
+static void
+create_f_shell_window()
+{
+    zwp_fullscreen_shell_v1_present_surface(wl_data.fshell,
+                                            win_data.shell_context->wl_shell_data.wl_surface,
+                                            ZWP_FULLSCREEN_SHELL_V1_PRESENT_METHOD_DEFAULT,
+                                            NULL);
+
+    /* Configure the surface so that it respects the width and height
+        * environment variables */
+    configure_surface_geometry(0, 0);
+}
+
+static shell_functions f_shell_functions = {.exit_fullscreen = &f_shell_exit_fullscreen,
+                                            .destroy_shell = &destroy_f_shell,
+                                            .create_window = &create_f_shell_window,
+                                            .enter_fullscreen = &f_shell_enter_fullscreen};
+
+f_shell_data f_data = {.functions = &f_shell_functions};
+
+//default executed when no shell exist
+static void
+no_shell_enter_fullscreen()
+{
+    g_warning("No available shell capable of fullscreening.");
+    win_data.is_fullscreen = false;
+}
+static void
+no_shell_exit_fullscreen()
+{
+    g_assert_not_reached();
+}
+static void
+maximize_no_shell()
```

Comment:
To be consistent with what you've done above, this should be called `no_shell_maximize`.